pallets/manta-pay/src/lib.rs

```diff
@@ -767,102 +788,172 @@ impl<L, R> AsRef<R> for WrapPair<L, R> {
         I: IntoIterator<Item = (Self::ValidUtxoAccumulatorOutput, Self::ValidNullifier)>,
     {
         let _ = super_key;
         let index = NullifierSetSize::<T>::get();
         let mut i = 0;
         for (_, nullifier) in iter {
-            let nullifier_commitment =
-                fp_encode(nullifier.0.nullifier.commitment).expect(FP_ENCODE);
+            let nullifier_commitment = fp_encode(nullifier.0.nullifier.commitment)
+                .map_err(SenderLedgerError::FpEncodeError)?;
             NullifierCommitmentSet::<T>::insert(nullifier_commitment, ());
             NullifierSetInsertionOrder::<T>::insert(
                 index + i,
                 (
                     nullifier_commitment,
                     OutgoingNote::try_from(nullifier.0.outgoing_note)
-                        .expect("Unable to decode the Outgoing Note."),
+                        .map_err(SenderLedgerError::OutgoingNoteDecodeError)?,
                 ),
             );
             i += 1;
         }
         if i != 0 {
             NullifierSetSize::<T>::set(index + i);
         }
+        Ok(())
     }
 }
 
+/// Merkle Tree Parameters Decode Error Type
+pub type MTParametersError = codec::DecodeError<
+    <&'static [u8] as codec::Read>::Error,
+    <config::UtxoAccumulatorModel as codec::Decode>::Error,
+>;
+
+/// Utxo Accumulator Item Hash Decode Error Type
+pub type UtxoItemHashError = codec::DecodeError<
+    <&'static [u8] as codec::Read>::Error,
+    <config::utxo::UtxoAccumulatorItemHash as codec::Decode>::Error,
+>;
+
+/// Receiver Ledger Error
+pub enum ReceiverLedgerError {
+    /// Utxo Decoding Error
+    UtxoDecodeError(scale_codec::Error),
+
+    /// Wrong Checksum Error
+    ChecksumError,
+
+    /// Merkle Tree Parameters Decoding Error
+    MTParametersDecodeError(MTParametersError),
+
+    /// Utxo Accumulator Item Hash Decoding Error
+    UtxoAccumulatorItemHashDecodeError(UtxoItemHashError),
+
+    /// Merkle Tree Out of Capacity Error
+    MerkleTreeCapacityError,
+
+    /// Field Element Encoding Error
+    FpEncodeError(scale_codec::Error),
+
+    /// Field Element Encoding Error
+    FpDecodeError(scale_codec::Error),
+
+    /// Path Decoding Error
+    PathDecodeError(scale_codec::Error),
+
+    /// Full Incoming Note Decoding Error
+    FullNoteDecodeError(scale_codec::Error),
+}
+
 impl<T> ReceiverLedger<config::Parameters> for Ledger<T>
 where
     T: Config,
 {
     type SuperPostingKey = (Wrap<()>, ());
     type ValidUtxo = Wrap<config::Utxo>;
+    type Error = ReceiverLedgerError;
 
     #[inline]
-    fn is_not_registered(&self, utxo: config::Utxo) -> Option<Self::ValidUtxo> {
-        if UtxoSet::<T>::contains_key(Utxo::try_from(utxo).ok()?) {
-            None
+    fn is_not_registered(
+        &self,
+        utxo: config::Utxo,
+    ) -> Result<Option<Self::ValidUtxo>, Self::Error> {
+        if UtxoSet::<T>::contains_key(
+            Utxo::try_from(utxo).map_err(ReceiverLedgerError::UtxoDecodeError)?,
```

Comment:
This is not really a decode error. The utxo can simply be missing.

Reply:
If Utxo is missing from storage then it returns `Ok(None)`

Which goes back to your the comment on that maybe should be an error

Reply:
oh wait my bad, this is the try_from error